### ISSUE TITLE
Rails 7+ / Ruby 3+  defined?(cookies) returns nil in view context

### ISSUE THREAD

**Issue by Valued** [rails/rails#49877]
Hello!
I faced this behavior on the working app after upgrading Ruby to 3.1 and Rails to 7.0.7.
Initially, I thought it could be something with configs, patches, etc. of the app but found nothing.
I created a brand-new app and was able to reproduce the issue. 

### Steps to reproduce

- add `debugger` to controller
- add `<% debugger %>` to view
- run the server, open the view and check the debugger console

### Expected behavior
Expected console output
Controller/View
```ruby
(ruby) defined? cookies
"method"
(rdbg) cookies
#<ActionDispatch::Cookies::CookieJar:...
 @committed=false,
 @cookies=
...
```

### Actual behavior
Actual console output
Controller
```ruby
(ruby) defined? cookies
"method"
(rdbg) cookies
#<ActionDispatch::Cookies::CookieJar:0x0000000111df6878
 @committed=false,
 @cookies=
...
(ruby) cookies.object_id
33860
```
View 
```ruby
(ruby) defined? cookies
nil
(rdbg) cookies
#<ActionDispatch::Cookies::CookieJar:0x0000000111df6878
 @committed=false,
 @cookies=
...
(ruby) cookies.object_id
33860
```
Note, that cookies are the same in both controller and view.

### System configuration
**Rails version**: 7.0.8

**Ruby version**: 3.2.2

**Comment by fatkodima** [comment]
Why do you use `defined?(cookies)` in the view instead of just `cookies`?

**Comment by Valued** [comment]
It looks like this
```
cookies[key] = value if defined?(cookies)

```
It's a part of a bigger code snippet that could be called from different parts of the app.
Author of this snippet added this check for safety reasons, in particular to avoid getting `NameError` if `cookies` aren't defined.

**Comment by fatkodima** [comment]
In which ruby and rails versions it worked before?

**Comment by Valued** [comment]
Ruby 2.7.4, Rails 6.0.6.1
Ruby 2.7.4, Rails 6.1.7.3

**Comment by fatkodima** [comment]
From what I explored, this is not dependent on a rails version, because it works with all rails versions when on ruby 2.7.x, for example. And fails for all versions when on ruby 3.1 (and 3.2).

**Comment by nvasilevski** [comment]
I think we can narrow it down to a small reproduction that points to a behavior change (or potentially a bug) in ruby

Here is a simplified version of what is going on:

```ruby
class Controller
  private

  def cookies
    {a_am_a: "cookie"}
  end
end

class View
  def initialize
    @controller = Controller.new
  end

  # cookies method is delegated to the controller
  def cookies
    @controller.cookies
  end

  def check_defined
    # this is what we do in the view
    puts defined?(cookies).inspect
  end

  def respond_to?(method, _bool)
    # this doesn't check for the private methods
    @controller.respond_to?(method)
  end
end

View.new.check_defined
```


But if we strip out all the Rails-like methods it will boil down to:
```ruby
class A
  def check_defined
    puts defined?(cookies).inspect
  end

  def respond_to?(_method, _bool)
    false
  end

  def cookies
  end
end

A.new.check_defined
```

It looks like in newer versions of ruby `defined?` takes into account the result of the `respond_to?` method. I'm going try to find whether it was an intentional change but meanwhile perhaps Rails should also consider changing this implementation:
https://github.com/rails/rails/blob/004d0a15234422beb34202c88dda27f46d6dd612/actionview/lib/action_view/helpers/controller_helper.rb#L37

to pass `true` to `respond_to?` to include private methods. Or maybe we should explicitly return `true` for the `cookies` method as its defined as `helper_method`
https://github.com/rails/rails/blob/004d0a15234422beb34202c88dda27f46d6dd612/actionpack/lib/action_controller/metal/cookies.rb#L8


I think there are two things we need to look into separately:
1. Was change to `defined?` behavior intentional?
2. `controller.respond_to?(:cookies)` should return `true` when its defined as a `helper_method`

**Comment by nvasilevski** [comment]
So this is where ruby started (from 3.0) using the result of `respond_to?` to determine the result of `defined?`
https://github.com/ruby/ruby/commit/184f78314e98cab63e7503cead4a4e99bd132a08

I don't think it's a ruby bug which means we have following options:
1. Make sure to forward `include_private` to `controller.respond_to?(method_name)` so `defined?` can keep looking for private methods as well
2. Do not delegate the `respond_to?` check to the controller if `self` (super)` responds to the method. And `self` responds to `cookies` as its defined as a helper method. 

Can't immediately tell which one I prefer

**Comment by fatkodima** [comment]
Good job! 👏 👏  I would probably choose 2nd solution.

**Comment by mguan2020** [comment]
I agree; the second solution avoids unnecessarily calling the controller's respond_to? method.

**Comment by Valued** [comment]
@fatkodima, @nvasilevski, @mguan2020, 
Thank you all so much for digging into this!
